File: pkg/dev_compiler/test/codegen_expected/corelib/shuffle_test.js
dart_library.library('corelib/shuffle_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__shuffle_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const _interceptors = dart_sdk._interceptors;
  const typed_data = dart_sdk.typed_data;
  const math = dart_sdk.math;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const shuffle_test = Object.create(null);
  let JSArrayOfint = () => (JSArrayOfint = dart.constFn(_interceptors.JSArray$(core.int)))();
  let ListOfint = () => (ListOfint = dart.constFn(core.List$(core.int)))();
  let intToint = () => (intToint = dart.constFn(dart.definiteFunctionType(core.int, [core.int])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  let VoidToint = () => (VoidToint = dart.constFn(dart.definiteFunctionType(core.int, [])))();
  let VoidTobool = () => (VoidTobool = dart.constFn(dart.definiteFunctionType(core.bool, [])))();
  let dynamicTovoid = () => (dynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic])))();
  shuffle_test.main = function() {
    for (let size of JSArrayOfint().of([0, 1, 2, 3, 7, 15, 99, 1023])) {
      let numbers = ListOfint().generate(size, dart.fn(x => x, intToint()));
      shuffle_test.testShuffle(numbers[dartx.toList]({growable: true}));
      shuffle_test.testShuffle(numbers[dartx.toList]({growable: false}));
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Uint32List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Int32List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Uint16List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Int16List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Uint8List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
      shuffle_test.testShuffle((() => {
        let _ = typed_data.Int8List.new(size);
        _[dartx.setAll](0, numbers);
        return _;
      })());
    }
    let l = JSArrayOfint().of([1, 2]);
    success: {
      for (let i = 0; i < 266; i++) {
        let first = core.int._check(l[dartx.first]);
        l[dartx.shuffle]();
        if (dart.equals(l[dartx.first], first)) break success;
      }
      expect$.Expect.fail("List changes every time.");
    }
    shuffle_test.testRandom();
  };
  dart.fn(shuffle_test.main, VoidTodynamic());
  shuffle_test.testShuffle = function(list) {
    let copy = core.List._check(dart.dsend(list, 'toList'));
    dart.dsend(list, 'shuffle');
    if (dart.test(dart.dsend(dart.dload(list, 'length'), '<', 2))) {
      expect$.Expect.listEquals(copy, core.List._check(list));
      return;
    }
    let seen = dart.map();
    for (let e of core.Iterable._check(list)) {
      seen[dartx.set](e, dart.dsend(seen[dartx.putIfAbsent](e, dart.fn(() => 0, VoidToint())), '+', 1));
    }
    for (let e of copy) {
      let remaining = core.int._check(seen[dartx.get](e));
      remaining = dart.notNull(remaining) - 1;
      if (remaining == 0) {
        seen[dartx.remove](e);
      } else {
        seen[dartx.set](e, remaining);
      }
    }
    expect$.Expect.isTrue(seen[dartx.isEmpty]);
    function listsDifferent() {
      for (let i = 0; i < dart.notNull(core.num._check(dart.dload(list, 'length'))); i++) {
        if (!dart.equals(dart.dindex(list, i), copy[dartx.get](i))) return true;
      }
      return false;
    }
    dart.fn(listsDifferent, VoidTobool());
    if (dart.test(dart.dsend(dart.dload(list, 'length'), '<', 59))) {
      let limit = 1e+80;
      let fact = 1.0;
      for (let i = 2; i < dart.notNull(core.num._check(dart.dload(list, 'length'))); i++) {
        fact = fact * i;
      }
      let combos = fact;
      while (!dart.test(listsDifferent()) && combos < limit) {
        dart.dsend(list, 'shuffle');
        combos = combos * fact;
      }
    }
    if (!dart.test(listsDifferent())) {
      expect$.Expect.fail(dart.str`Didn't shuffle at all, p < 1:1e80: ${list}`);
    }
  };
  dart.fn(shuffle_test.testShuffle, dynamicTovoid());
  shuffle_test.testRandom = function() {
    let randomNums = JSArrayOfint().of([37, 87, 42, 157, 252, 17]);
    let numbers = core.List.generate(25, dart.fn(x => x, intToint()));
    let l1 = numbers[dartx.toList]();
    l1[dartx.shuffle](new shuffle_test.MockRandom(ListOfint()._check(randomNums)));
    for (let i = 0; i < 50; i++) {
      let l2 = numbers[dartx.toList]();
      l2[dartx.shuffle](new shuffle_test.MockRandom(ListOfint()._check(randomNums)));
      expect$.Expect.listEquals(l1, l2);
    }
  };
  dart.fn(shuffle_test.testRandom, VoidTodynamic());
  const _values = Symbol('_values');
  const _next = Symbol('_next');
  shuffle_test.MockRandom = class MockRandom extends core.Object {
    new(values) {
      this[_values] = values;
      this.index = 0;
    }
    get [_next]() {
      let next = this[_values][dartx.get](this.index);
      this.index = (dart.notNull(this.index) + 1)[dartx['%']](this[_values][dartx.length]);
      return next;
    }
    nextInt(limit) {
      return this[_next][dartx['%']](limit);
    }
    nextDouble() {
      return dart.notNull(this[_next]) / 256.0;
    }
    nextBool() {
      return this[_next][dartx.isEven];
    }
  };
  shuffle_test.MockRandom[dart.implements] = () => [math.Random];
  dart.setSignature(shuffle_test.MockRandom, {
    constructors: () => ({new: dart.definiteFunctionType(shuffle_test.MockRandom, [core.List$(core.int)])}),
    methods: () => ({
      nextInt: dart.definiteFunctionType(core.int, [core.int]),
      nextDouble: dart.definiteFunctionType(core.double, []),
      nextBool: dart.definiteFunctionType(core.bool, [])
    })
  });
  // Exports:
  exports.shuffle_test = shuffle_test;
});

File: tests/corelib/shuffle_test.dart
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// Dart test for List.shuffle.
library shuffle_test;

import "dart:typed_data";
import "dart:math" show Random;
import "package:expect/expect.dart";

main() {
  for (int size in [0, 1, 2, 3, 7, 15, 99, 1023]) {
    var numbers = new List<int>.generate(size, (x) => x);
    testShuffle(numbers.toList(growable: true));
    testShuffle(numbers.toList(growable: false));
    testShuffle(new Uint32List(size)..setAll(0, numbers));
    testShuffle(new Int32List(size)..setAll(0, numbers));
    testShuffle(new Uint16List(size)..setAll(0, numbers));
    testShuffle(new Int16List(size)..setAll(0, numbers));
    // Some numbers will be truncated in the following two.
    testShuffle(new Uint8List(size)..setAll(0, numbers));
    testShuffle(new Int8List(size)..setAll(0, numbers));
    //testShuffle(numbers.map((x) => "$x").toList());
  }

  // Check that it actually can keep the same list (regression test).
  List l = [1, 2];
  success:
  {
    for (int i = 0; i < 266; i++) {
      int first = l.first;
      l.shuffle();
      if (l.first == first) break success; // List didn't change.
    }
    // Chance of changing 266 times in a row should be < 1:1e80.
    Expect.fail("List changes every time.");
  }

  testRandom();
}

void testShuffle(list) {
  List copy = list.toList();
  list.shuffle();
  if (list.length < 2) {
    Expect.listEquals(copy, list);
    return;
  }
  // Test that the list after shuffling has the same elements as before,
  // without considering order.
  Map seen = {};
  for (var e in list) {
    seen[e] = seen.putIfAbsent(e, () => 0) + 1;
  }
  for (var e in copy) {
    int remaining = seen[e];
    remaining -= 1; // Throws if e was not in map at all.
    if (remaining == 0) {
      seen.remove(e);
    } else {
      seen[e] = remaining;
    }
  }
  Expect.isTrue(seen.isEmpty);
  // Test that shuffle actually does make a change. Repeat until the probability
  // of a proper shuffling hitting the same list again is less than 10^80
  // (arbitrary bignum - approx. number of atoms in the universe).
  //
  // The probablility of shuffling a list of length n into the same list is
  // 1/n!. If one shuffle didn't change the list, repeat shuffling until
  // probability of randomly hitting the same list every time is less than
  // 1/1e80.

  bool listsDifferent() {
    for (int i = 0; i < list.length; i++) {
      if (list[i] != copy[i]) return true;
    }
    return false;
  }

  if (list.length < 59) {
    // 59! > 1e80.
    double limit = 1e80;
    double fact = 1.0;
    for (int i = 2; i < list.length; i++) fact *= i;
    double combos = fact;

    while (!listsDifferent() && combos < limit) {
      list.shuffle();
      combos *= fact;
    }
  }
  if (!listsDifferent()) {
    Expect.fail("Didn't shuffle at all, p < 1:1e80: $list");
  }
}

// Checks that the "random" argument to shuffle is used.
testRandom() {
  List<int> randomNums = [37, 87, 42, 157, 252, 17];
  List numbers = new List.generate(25, (x) => x);
  List l1 = numbers.toList()..shuffle(new MockRandom(randomNums));
  for (int i = 0; i < 50; i++) {
    // With same random sequence, we get the same shuffling each time.
    List l2 = numbers.toList()..shuffle(new MockRandom(randomNums));
    Expect.listEquals(l1, l2);
  }
}

class MockRandom implements Random {
  final List<int> _values;
  int index = 0;
  MockRandom(this._values);

  int get _next {
    int next = _values[index];
    index = (index + 1) % _values.length;
    return next;
  }

  int nextInt(int limit) => _next % limit;

  double nextDouble() => _next / 256.0;

  bool nextBool() => _next.isEven;
}
